File: lib/repolist.js
var winston = require('./winston');
var bluebird = require('bluebird');
var check = require('check-types');

module.exports = function(github) {
  check.assert.object(github);
  this.github = github;
}

module.exports.prototype = {
  getRepos: function() {
    var self = this;
    return self.github.user.getTeamsAsync({}).then(function(teams) {
      var p = [];
      teams.forEach(function(team) {
        winston.info("Looking for all repos in team %s/%s",
            team.organization.login,
            team.name);
        p.push(self.github.orgs.getTeamReposAsync({
          id: team.id,
          per_page: 100
        }));
      });
      return bluebird.all(p);
    }).then(function(repoSets) {
      var p = [];
      repoSets.forEach(function(set) {
        set.forEach(function(repo) {
          p.push(repo);
        });
      });
      return p;
    });
  },
  updateHooks: function() {
    var self = this;
    return self.getRepos().each(function(repo) {
      return self.github.repos.getHooksAsync({
        user: repo.owner.login,
        repo: repo.name,
      }).then(function(hooks) {
        var active = false;
        hooks.forEach(function(hook) {
          if (hook.config.url) {
            if (hook.config.url.indexOf("http://build.codius.org/github-hook") === 0) {
              active = true;
            }
          }
        });
        if (!active) {
          winston.info("Enabling hook on %s/%s", repo.owner.login, repo.name);
          return self.github.repos.createHookAsync({
            user: repo.owner.login,
            repo: repo.name,
            name: 'web',
            config: {
              url: 'http://build.codius.org/github-hook',
              content_type: 'json'
            },
            events: '*',
            active: true
          });
        }
      });
    });
  }
}

File: repolist.go
package monty

import (
	"fmt"
	"github.com/google/go-github/github"
)

type RepositoryList struct {
	client *github.Client
}

func NewRepolist(client *github.Client) *RepositoryList {
	ret := new(RepositoryList)
	ret.client = client
	return ret
}

func (self *RepositoryList) List() *[]Repo {
	var repolist = make([]Repo, 0)

	teams, _, _ := self.client.Organizations.ListUserTeams(nil)
	for _, team := range teams {
		repos, _, _ := self.client.Organizations.ListTeamRepos(*team.ID, nil)
		for _, repo := range repos {
			repolist = append(repolist, Repo{
				FullName: repo.FullName,
				Owner:    repo.Owner.Login,
				Name:     repo.Name,
			})
		}
	}

	return &repolist
}

func (self *RepositoryList) EnableHooks() {
	for _, repo := range *self.List() {
		self.EnableHook(&repo)
	}
}

func (self *RepositoryList) EnableHook(repo *Repo) {
	hooks, _, _ := self.client.Repositories.ListHooks(*repo.Owner, *repo.Name, nil)

	fmt.Println("Checking hooks on", *repo.FullName)
	hasHook := false

	for _, hook := range hooks {
		if *hook.Name == "web" {
			fmt.Println("\tHook:", hook.Config["url"])
			if hook.Config["url"] == "http://build.codius.org/github-hook" {
				hasHook = true
			}
		}
	}

	if !hasHook {
		fmt.Println("\tNeeds a hook!")
		name := "web"
		hook := github.Hook{
			Name: &name,
			Config: map[string]interface{}{
				"url": "http://build.codius.org/github-hook",
			},
			Events: []string{"*"},
		}
		self.client.Repositories.CreateHook(*repo.Owner, *repo.Name, &hook)
	}
}
